File: sdks/ts/memory/src/ingest/extract/video.test.ts
// SPDX-License-Identifier: Apache-2.0

import { execFileSync } from 'node:child_process'
import { readFileSync } from 'node:fs'
import { mkdtempSync, rmSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { Readable } from 'node:stream'
import { afterEach, beforeEach, describe, expect, it } from 'vitest'

import { AudioExtractor } from './audio.js'
import { resetBinaryCache } from './subprocess.js'
import {
  VideoExtractor,
  type VideoExtractorConfig,
  type VideoMetadata,
  buildVideoMetadata,
  computeExtractionTimeout,
  mergeMetadata,
  parseFrameRate,
} from './video.js'

// ---- Helpers ----

const hasFFmpeg = (): boolean => {
  try {
    execFileSync('which', ['ffmpeg'], { stdio: 'ignore' })
    return true
  } catch {
    return false
  }
}

const hasFFprobe = (): boolean => {
  try {
    execFileSync('which', ['ffprobe'], { stdio: 'ignore' })
    return true
  } catch {
    return false
  }
}

const generateTestVideo = (outputPath: string, withAudio: boolean): void => {
  const args = [
    '-y',
    '-f', 'lavfi', '-i', 'testsrc=duration=1:size=160x120:rate=10',
    ...(withAudio ? ['-f', 'lavfi', '-i', 'sine=frequency=440:duration=1'] : []),
    '-c:v', 'libx264', '-preset', 'ultrafast',
    ...(withAudio ? ['-c:a', 'aac', '-shortest'] : ['-an']),
    outputPath,
  ]
  execFileSync('ffmpeg', args, { stdio: 'ignore' })
}

const makeExtractor = (overrides?: Partial<VideoExtractorConfig>): VideoExtractor =>
  new VideoExtractor({
    audioExtractor: new AudioExtractor(),
    ...overrides,
  })

// ---- Unit tests ----

describe('VideoExtractor', () => {
  beforeEach(() => {
    resetBinaryCache()
  })

  afterEach(() => {
    resetBinaryCache()
  })

  it('has the correct name', () => {
    const e = makeExtractor()
    expect(e.name).toBe('video-ffmpeg')
  })

  describe('capability', () => {
    it('includes expected extensions', () => {
      const e = makeExtractor()
      const cap = e.capability()
      const expected = ['.mp4', '.avi', '.mkv', '.mov', '.webm', '.wmv', '.flv', '.m4v']
      for (const ext of expected) {
        expect(cap.extensions).toContain(ext)
      }
    })

    it('includes expected MIME types', () => {
      const e = makeExtractor()
      const cap = e.capability()
      const expected = [
        'video/mp4',
        'video/x-msvideo',
        'video/x-matroska',
        'video/quicktime',
        'video/webm',
        'video/x-ms-wmv',
        'video/x-flv',
      ]
      for (const mime of expected) {
        expect(cap.mimeTypes).toContain(mime)
      }
    })

    it('requires binary', () => {
      const e = makeExtractor()
      expect(e.capability().requiresBinary).toBe(true)
    })

    it('has magic byte signatures', () => {
      const e = makeExtractor()
      const cap = e.capability()
      expect(cap.magicBytes).toBeDefined()
      expect(cap.magicBytes.length).toBeGreaterThan(0)
    })
  })

  describe('available', () => {
    it('returns false when ffmpeg is absent', async () => {
      const e = makeExtractor({ ffmpegBinary: '/nonexistent/ffmpeg' })
      const result = await e.available()
      expect(result).toBe(false)
    })

    it('returns false when ffprobe is absent', async () => {
      const e = makeExtractor({ ffprobeBinary: '/nonexistent/ffprobe' })
      const result = await e.available()
      expect(result).toBe(false)
    })
  })

  describe('extract', () => {
    it('rejects files exceeding the size limit', async () => {
      const e = makeExtractor({ maxFileSizeBytes: 1024 })
      const input = Buffer.alloc(2048)
      await expect(e.extract(input, {})).rejects.toThrow('exceeds maximum')
    })
  })

  describe('extractStream', () => {
    it('rejects streams exceeding the size limit', async () => {
      const e = makeExtractor({ maxFileSizeBytes: 1024 })
      const input = Readable.from(Buffer.alloc(2048))
      await expect(e.extractStream(input, {})).rejects.toThrow('exceeds maximum')
    })
  })
})

// ---- Pure function tests ----

describe('parseFrameRate', () => {
  it('parses standard frame rates', () => {
    expect(parseFrameRate('30/1')).toBe(30)
    expect(parseFrameRate('60/1')).toBe(60)
  })

  it('parses fractional frame rates', () => {
    const result = parseFrameRate('24000/1001')
    expect(result).toBeDefined()
    expect(result!).toBeCloseTo(23.976, 2)
  })

  it('returns undefined for invalid input', () => {
    expect(parseFrameRate(undefined)).toBeUndefined()
    expect(parseFrameRate('')).toBeUndefined()
    expect(parseFrameRate('invalid')).toBeUndefined()
    expect(parseFrameRate('abc/def')).toBeUndefined()
  })

  it('returns undefined for division by zero', () => {
    expect(parseFrameRate('30/0')).toBeUndefined()
  })
})

describe('buildVideoMetadata', () => {
  it('builds metadata from complete ffprobe output', () => {
    const probe = {
      format: { duration: '125.4' },
      streams: [
        {
          codec_type: 'video',
          codec_name: 'h264',
          width: 1920,
          height: 1080,
          r_frame_rate: '30/1',
        },
        {
          codec_type: 'audio',
          codec_name: 'aac',
        },
      ],
    }

    const meta = buildVideoMetadata(probe)
    expect(meta.duration_seconds).toBe(125.4)
    expect(meta.width).toBe(1920)
    expect(meta.height).toBe(1080)
    expect(meta.codec).toBe('h264')
    expect(meta.has_audio).toBe(true)
    expect(meta.frame_rate).toBe(30)
  })

  it('detects no audio stream', () => {
    const probe = {
      format: { duration: '60.0' },
      streams: [
        {
          codec_type: 'video',
          codec_name: 'vp9',
          width: 1280,
          height: 720,
          r_frame_rate: '24000/1001',
        },
      ],
    }

    const meta = buildVideoMetadata(probe)
    expect(meta.has_audio).toBe(false)
  })

  it('handles empty duration', () => {
    const probe = { format: {}, streams: [] }
    const meta = buildVideoMetadata(probe)
    expect(meta.duration_seconds).toBe(0)
  })

  it('handles invalid duration', () => {
    const probe = { format: { duration: 'invalid' }, streams: [] }
    const meta = buildVideoMetadata(probe)
    expect(meta.duration_seconds).toBe(0)
  })
})

describe('computeExtractionTimeout', () => {
  it('returns at least 60 seconds for short videos', () => {
    const timeout = computeExtractionTimeout(30, 120_000)
    expect(timeout).toBeGreaterThanOrEqual(60_000)
  })

  it('scales with video duration', () => {
    const short = computeExtractionTimeout(300, 120_000)
    const long = computeExtractionTimeout(3600, 120_000)
    expect(long).toBeGreaterThan(short)
  })

  it('computes proportional timeout for 20-minute video', () => {
    const timeout = computeExtractionTimeout(1200, 120_000)
    expect(timeout).toBeGreaterThanOrEqual(240_000)
  })
})

describe('mergeMetadata', () => {
  it('combines video and audio metadata', () => {
    const video: VideoMetadata = {
      duration_seconds: 120,
      width: 1920,
      height: 1080,
      has_audio: true,
    }
    const audio: Record<string, string> = {
      detected_language: 'en',
      segments: '[{"start":0,"end":5,"text":"hello"}]',
    }

    const merged = mergeMetadata(video, audio)
    expect(merged['detected_language']).toBe('en')
    expect(merged['video_duration_seconds']).toBe('120.00')
    expect(merged['video_width']).toBe('1920')
    expect(merged['video_height']).toBe('1080')
    expect(merged['video_has_audio']).toBe('true')
    expect(merged['segments']).toBeDefined()
  })

  it('preserves all audio metadata keys', () => {
    const video: VideoMetadata = {
      duration_seconds: 60,
      has_audio: true,
    }
    const audio: Record<string, string> = {
      key1: 'value1',
      key2: '42',
      key3: 'true',
    }

    const merged = mergeMetadata(video, audio)
    expect(merged['key1']).toBe('value1')
    expect(merged['key2']).toBe('42')
    expect(merged['key3']).toBe('true')
  })
})

// ---- Integration tests (require ffmpeg/ffprobe) ----

describe('VideoExtractor integration', () => {
  let tmpDir: string

  beforeEach(() => {
    tmpDir = mkdtempSync(join(tmpdir(), 'memory-video-test-'))
  })

  afterEach(() => {
    rmSync(tmpDir, { recursive: true, force: true })
  })

  it.skipIf(!hasFFmpeg() || !hasFFprobe())(
    'verifies video data can be read for extraction',
    async () => {
      const videoPath = join(tmpDir, 'test.mp4')
      generateTestVideo(videoPath, true)
      const videoData = readFileSync(videoPath)
      expect(videoData.length).toBeGreaterThan(0)
    },
  )

  it.skipIf(!hasFFmpeg() || !hasFFprobe())(
    'handles video with no audio track gracefully',
    async () => {
      const videoPath = join(tmpDir, 'silent.mp4')
      generateTestVideo(videoPath, false)

      const videoData = readFileSync(videoPath)
      const e = makeExtractor()
      const result = await e.extract(videoData, {})

      expect(result.text).toContain('no audio track')
      expect(result.metadata['video_has_audio']).toBe('false')
      expect(result.metadata['video_duration_seconds']).toBeDefined()
    },
  )

  it.skipIf(!hasFFmpeg() || !hasFFprobe())(
    'handles video with no audio track via streaming',
    async () => {
      const videoPath = join(tmpDir, 'silent-stream.mp4')
      generateTestVideo(videoPath, false)

      const videoData = readFileSync(videoPath)
      const stream = Readable.from(videoData)
      const e = makeExtractor()
      const result = await e.extractStream(stream, {})

      expect(result.text).toContain('no audio track')
    },
  )
})

describe('VideoExtractor config', () => {
  it('uses default config values', () => {
    const e = makeExtractor()
    expect(e.name).toBe('video-ffmpeg')
  })

  it('disables keyframe extraction by default', () => {
    const e = makeExtractor()
    expect(e.name).toBe('video-ffmpeg')
  })
})

// ---- Mock-based pipeline test (MA-1) ----

describe('VideoExtractor mock pipeline', () => {
  it('exercises probe-parse and merge logic without FFmpeg', () => {
    // Simulate FFprobe JSON output parsing.
    const fakeProbeOutput = {
      format: { duration: '90.5' },
      streams: [
        {
          codec_type: 'video',
          codec_name: 'h264',
          width: 1920,
          height: 1080,
          r_frame_rate: '30/1',
        },
        {
          codec_type: 'audio',
          codec_name: 'aac',
        },
      ],
    }

    const videoMeta = buildVideoMetadata(fakeProbeOutput)
    expect(videoMeta.duration_seconds).toBe(90.5)
    expect(videoMeta.has_audio).toBe(true)
    expect(videoMeta.width).toBe(1920)
    expect(videoMeta.height).toBe(1080)
    expect(videoMeta.codec).toBe('h264')
    expect(videoMeta.frame_rate).toBe(30)

    // Simulate AudioExtractor result from transcription.
    const fakeAudioResult: Record<string, string> = {
      detected_language: 'en',
      segments: '[{"start":0,"end":30,"text":"Hello world."}]',
      duration_seconds: '30.0',
    }

    const merged = mergeMetadata(videoMeta, fakeAudioResult)
    merged['source_bytes'] = '1024'

    expect(merged['detected_language']).toBe('en')
    expect(merged['video_duration_seconds']).toBe('90.50')
    expect(merged['source_bytes']).toBe('1024')
    expect(merged['segments']).toBeDefined()
  })

  it('returns no-audio result when probe shows no audio stream', () => {
    const noAudioProbe = {
      format: { duration: '10.0' },
      streams: [
        {
          codec_type: 'video',
          codec_name: 'vp9',
          width: 640,
          height: 480,
          r_frame_rate: '25/1',
        },
      ],
    }

    const meta = buildVideoMetadata(noAudioProbe)
    expect(meta.has_audio).toBe(false)

    // The extractor would return this message for the no-audio path.
    const result = {
      text: 'Video contains no audio track.',
      contentType: 'text/plain',
      metadata: { video_has_audio: 'false' },
    }
    expect(result.text).toContain('no audio track')
    expect(result.metadata['video_has_audio']).toBe('false')
  })
})

// ---- Corrupt/invalid video test (MA-2) ----

describe('VideoExtractor corrupt video', () => {
  let tmpDir: string

  beforeEach(() => {
    tmpDir = mkdtempSync(join(tmpdir(), 'memory-video-corrupt-'))
  })

  afterEach(() => {
    rmSync(tmpDir, { recursive: true, force: true })
  })

  it.skipIf(!hasFFmpeg() || !hasFFprobe())(
    'produces a meaningful error for corrupt video input',
    async () => {
      // Write garbage with ftyp magic bytes to pass magic-byte checks.
      const garbage = Buffer.alloc(512)
      garbage.write('ftyp', 4, 'ascii')

      const e = makeExtractor()
      await expect(e.extract(garbage, {})).rejects.toThrow(/ffprobe|ffmpeg/i)
    },
  )

  it.skipIf(!hasFFmpeg() || !hasFFprobe())(
    'produces a meaningful error for corrupt video via streaming',
    async () => {
      const garbage = Buffer.alloc(512)
      garbage.write('ftyp', 4, 'ascii')

      const stream = Readable.from(garbage)
      const e = makeExtractor()
      await expect(e.extractStream(stream, {})).rejects.toThrow(/ffprobe|ffmpeg/i)
    },
  )
})

// ---- Streaming size-limit enforcement during copy (M-3) ----

describe('writeStreamToFile size enforcement', () => {
  let tmpDir: string

  beforeEach(() => {
    tmpDir = mkdtempSync(join(tmpdir(), 'memory-video-stream-'))
  })

  afterEach(() => {
    rmSync(tmpDir, { recursive: true, force: true })
  })

  it('rejects a stream that gradually exceeds the limit during copy', async () => {
    // The maxFileSizeBytes is set low and the stream delivers more data
    // than the limit. This tests enforcement during the actual copy, not
    // just the pre-check.
    const e = makeExtractor({ maxFileSizeBytes: 64 })
    const oversizedStream = Readable.from(Buffer.alloc(128))
    await expect(
      e.extractStream(oversizedStream, {}),
    ).rejects.toThrow('exceeds maximum')
  })
})

File: go/ingest/video_test.go
// SPDX-License-Identifier: Apache-2.0

package ingest

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

func TestVideoExtractor_Name(t *testing.T) {
	e := NewVideoExtractor(VideoExtractorConfig{})
	if got := e.Name(); got != "video-ffmpeg" {
		t.Errorf("Name() = %q, want %q", got, "video-ffmpeg")
	}
}

func TestVideoExtractor_ContentTypes(t *testing.T) {
	e := NewVideoExtractor(VideoExtractorConfig{})
	types := e.ContentTypes()
	if len(types) == 0 {
		t.Fatal("ContentTypes() returned empty slice")
	}
	wantTypes := map[string]bool{
		"video/mp4":        true,
		"video/x-msvideo":  true,
		"video/x-matroska": true,
		"video/quicktime":  true,
		"video/webm":       true,
		"video/x-ms-wmv":   true,
		"video/x-flv":      true,
	}
	for _, ct := range types {
		if !wantTypes[ct] {
			t.Errorf("unexpected content type %q", ct)
		}
		delete(wantTypes, ct)
	}
	for ct := range wantTypes {
		t.Errorf("missing expected content type %q", ct)
	}
}

func TestVideoExtractor_Capability(t *testing.T) {
	e := NewVideoExtractor(VideoExtractorConfig{})
	cap := e.Capability()

	wantExts := map[string]bool{
		".mp4": true, ".avi": true, ".mkv": true, ".mov": true,
		".webm": true, ".wmv": true, ".flv": true, ".m4v": true,
	}
	for _, ext := range cap.Extensions {
		if !wantExts[ext] {
			t.Errorf("unexpected extension %q in capability", ext)
		}
		delete(wantExts, ext)
	}
	for ext := range wantExts {
		t.Errorf("missing expected extension %q in capability", ext)
	}

	if !cap.RequiresBinary {
		t.Error("RequiresBinary should be true")
	}

	if len(cap.MagicBytes) == 0 {
		t.Error("expected at least one magic byte signature")
	}
}

func TestVideoExtractor_Available_NoFFmpeg(t *testing.T) {
	ResetBinaryCache()
	defer ResetBinaryCache()

	e := NewVideoExtractor(VideoExtractorConfig{
		FFmpegBinary: "/nonexistent/ffmpeg",
	})
	ok, err := e.Available(t.Context())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if ok {
		t.Error("Available() should return false when ffmpeg is absent")
	}
}

func TestVideoExtractor_Available_NoFFprobe(t *testing.T) {
	ResetBinaryCache()
	defer ResetBinaryCache()

	e := NewVideoExtractor(VideoExtractorConfig{
		FFprobeBinary: "/nonexistent/ffprobe",
	})
	ok, err := e.Available(t.Context())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if ok {
		t.Error("Available() should return false when ffprobe is absent")
	}
}

func TestVideoExtractor_Available_NoAudioExtractor(t *testing.T) {
	e := NewVideoExtractor(VideoExtractorConfig{
		AudioExtractor: nil,
	})
	ok, err := e.Available(t.Context())
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if ok {
		t.Error("Available() should return false when AudioExtractor is nil")
	}
}

func TestVideoExtractor_Extract_FileTooLarge(t *testing.T) {
	e := NewVideoExtractor(VideoExtractorConfig{
		MaxFileSize: 1024,
	})

	input := make([]byte, 2048)
	_, err := e.Extract(t.Context(), input, ExtractOptions{})
	if err == nil {
		t.Fatal("expected error for oversized file")
	}
	if !strings.Contains(err.Error(), "exceeds maximum") {
		t.Errorf("error should mention size limit: %v", err)
	}
}

func TestVideoExtractor_Extract_NoAudioExtractorDependency(t *testing.T) {
	e := NewVideoExtractor(VideoExtractorConfig{
		AudioExtractor: nil,
	})
	_, err := e.Extract(t.Context(), []byte("dummy"), ExtractOptions{})
	if err == nil {
		t.Fatal("expected error when AudioExtractor is nil")
	}
	if !strings.Contains(err.Error(), "AudioExtractor") {
		t.Errorf("error should mention AudioExtractor: %v", err)
	}
}

func TestVideoExtractor_DefaultConfig(t *testing.T) {
	cfg := VideoExtractorConfig{}
	cfg.applyDefaults()

	if cfg.FFmpegBinary != "ffmpeg" {
		t.Errorf("default FFmpegBinary = %q, want %q", cfg.FFmpegBinary, "ffmpeg")
	}
	if cfg.FFprobeBinary != "ffprobe" {
		t.Errorf("default FFprobeBinary = %q, want %q", cfg.FFprobeBinary, "ffprobe")
	}
	if cfg.MaxFileSize != defaultMaxVideoSize {
		t.Errorf("default MaxFileSize = %d, want %d", cfg.MaxFileSize, defaultMaxVideoSize)
	}
	if cfg.ExtractionTimeout != defaultVideoExtractionTimeout {
		t.Errorf("default ExtractionTimeout = %v, want %v", cfg.ExtractionTimeout, defaultVideoExtractionTimeout)
	}
	if cfg.MaxKeyframes != defaultMaxKeyframes {
		t.Errorf("default MaxKeyframes = %d, want %d", cfg.MaxKeyframes, defaultMaxKeyframes)
	}
}

func TestVideoExtractor_EnvVarOverride(t *testing.T) {
	t.Setenv("MEMORY_FFMPEG_PATH", "/custom/ffmpeg")
	t.Setenv("MEMORY_FFPROBE_PATH", "/custom/ffprobe")

	cfg := VideoExtractorConfig{}
	cfg.applyDefaults()

	if cfg.FFmpegBinary != "/custom/ffmpeg" {
		t.Errorf("FFmpegBinary = %q, want %q", cfg.FFmpegBinary, "/custom/ffmpeg")
	}
	if cfg.FFprobeBinary != "/custom/ffprobe" {
		t.Errorf("FFprobeBinary = %q, want %q", cfg.FFprobeBinary, "/custom/ffprobe")
	}
}

func TestBuildVideoMetadata(t *testing.T) {
	probe := ffprobeOutput{
		Format: ffprobeFormat{Duration: "125.4"},
		Streams: []ffprobeStream{
			{
				CodecType:  "video",
				CodecName:  "h264",
				Width:      1920,
				Height:     1080,
				RFrameRate: "30/1",
			},
			{
				CodecType: "audio",
				CodecName: "aac",
			},
		},
	}

	meta := buildVideoMetadata(probe)

	if meta.DurationSeconds != 125.4 {
		t.Errorf("DurationSeconds = %f, want 125.4", meta.DurationSeconds)
	}
	if meta.Width != 1920 {
		t.Errorf("Width = %d, want 1920", meta.Width)
	}
	if meta.Height != 1080 {
		t.Errorf("Height = %d, want 1080", meta.Height)
	}
	if meta.Codec != "h264" {
		t.Errorf("Codec = %q, want %q", meta.Codec, "h264")
	}
	if !meta.HasAudio {
		t.Error("HasAudio should be true")
	}
	if meta.FrameRate != 30.0 {
		t.Errorf("FrameRate = %f, want 30.0", meta.FrameRate)
	}
}

func TestBuildVideoMetadata_NoAudio(t *testing.T) {
	probe := ffprobeOutput{
		Format: ffprobeFormat{Duration: "60.0"},
		Streams: []ffprobeStream{
			{
				CodecType:  "video",
				CodecName:  "vp9",
				Width:      1280,
				Height:     720,
				RFrameRate: "24000/1001",
			},
		},
	}

	meta := buildVideoMetadata(probe)
	if meta.HasAudio {
		t.Error("HasAudio should be false when no audio stream present")
	}
	if meta.FrameRate < 23.9 || meta.FrameRate > 24.0 {
		t.Errorf("FrameRate = %f, want ~23.976", meta.FrameRate)
	}
}

func TestBuildVideoMetadata_EmptyDuration(t *testing.T) {
	probe := ffprobeOutput{
		Format:  ffprobeFormat{Duration: ""},
		Streams: []ffprobeStream{},
	}
	meta := buildVideoMetadata(probe)
	if meta.DurationSeconds != 0 {
		t.Errorf("DurationSeconds = %f, want 0", meta.DurationSeconds)
	}
}

func TestParseFrameRate(t *testing.T) {
	tests := []struct {
		input string
		want  float64
	}{
		{"30/1", 30.0},
		{"24000/1001", 23.976},
		{"0/0", 0},
		{"invalid", 0},
		{"abc/def", 0},
		{"", 0},
	}

	for _, tt := range tests {
		got := parseFrameRate(tt.input)
		if tt.want == 0 && got != 0 {
			t.Errorf("parseFrameRate(%q) = %f, want 0", tt.input, got)
		}
		if tt.want != 0 {
			diff := got - tt.want
			if diff < 0 {
				diff = -diff
			}
			if diff > 0.01 {
				t.Errorf("parseFrameRate(%q) = %f, want ~%f", tt.input, got, tt.want)
			}
		}
	}
}

func TestComputeExtractionTimeout(t *testing.T) {
	e := NewVideoExtractor(VideoExtractorConfig{})

	// Short video: should get at least 60s.
	timeout := e.computeExtractionTimeout(30)
	if timeout < 60*time.Second {
		t.Errorf("short video timeout = %v, want >= 60s", timeout)
	}

	// 20-minute video: ~2 blocks of 10 min = 240s.
	timeout = e.computeExtractionTimeout(1200)
	if timeout < 200*time.Second {
		t.Errorf("20-min video timeout = %v, want >= 200s", timeout)
	}
}

func TestMergeVideoAudioMetadata(t *testing.T) {
	video := VideoMetadata{
		DurationSeconds: 120,
		Width:           1920,
		Height:          1080,
		HasAudio:        true,
	}
	audio := map[string]string{
		"detected_language": "en",
		"segment_count":     "5",
	}

	merged := mergeVideoAudioMetadata(video, audio)

	if merged["detected_language"] != "en" {
		t.Error("audio metadata not preserved")
	}
	if merged["video_duration_seconds"] != "120.00" {
		t.Errorf("video_duration_seconds = %q, want %q", merged["video_duration_seconds"], "120.00")
	}
	if merged["video_width"] != "1920" {
		t.Errorf("video_width = %q, want %q", merged["video_width"], "1920")
	}
	if merged["video_height"] != "1080" {
		t.Errorf("video_height = %q, want %q", merged["video_height"], "1080")
	}
	if merged["video_has_audio"] != "true" {
		t.Errorf("video_has_audio = %q, want %q", merged["video_has_audio"], "true")
	}
}

func TestWriteVideoToTempFile(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "test.bin")
	data := []byte("hello world")

	n, err := writeVideoToTempFile(path, bytes.NewReader(data), 1024)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if n != int64(len(data)) {
		t.Errorf("written = %d, want %d", n, len(data))
	}

	got, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(got, data) {
		t.Error("file content mismatch")
	}
}

func TestWriteVideoToTempFile_ExceedsLimit(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "test.bin")
	data := make([]byte, 100)

	_, err := writeVideoToTempFile(path, bytes.NewReader(data), 50)
	if err == nil {
		t.Fatal("expected error for oversized input")
	}
	if !strings.Contains(err.Error(), "exceeds maximum") {
		t.Errorf("error should mention limit: %v", err)
	}
}

func TestWriteVideoToTempFile_Permissions(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "perms.bin")
	data := []byte("secret video data")

	_, err := writeVideoToTempFile(path, bytes.NewReader(data), 1024)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	info, err := os.Stat(path)
	if err != nil {
		t.Fatal(err)
	}
	perm := info.Mode().Perm()
	if perm != 0o600 {
		t.Errorf("file permissions = %o, want 0600", perm)
	}
}

func TestVideoMetadata_JSONRoundTrip(t *testing.T) {
	meta := VideoMetadata{
		DurationSeconds: 123.45,
		Width:           1920,
		Height:          1080,
		Codec:           "h264",
		HasAudio:        true,
		FrameRate:       29.97,
	}

	data, err := json.Marshal(meta)
	if err != nil {
		t.Fatal(err)
	}

	var decoded VideoMetadata
	if err := json.Unmarshal(data, &decoded); err != nil {
		t.Fatal(err)
	}

	if decoded != meta {
		t.Errorf("round-trip mismatch: got %+v, want %+v", decoded, meta)
	}
}

func TestVideoExtractor_KeyframeExtractionDisabledByDefault(t *testing.T) {
	cfg := VideoExtractorConfig{}
	cfg.applyDefaults()
	if cfg.KeyframeExtraction {
		t.Error("KeyframeExtraction should be disabled by default")
	}
}

func TestVideoExtractor_OverrideTimeout(t *testing.T) {
	e := NewVideoExtractor(VideoExtractorConfig{})

	if got := e.overrideTimeout(); got != 0 {
		t.Errorf("overrideTimeout() without env = %v, want 0", got)
	}

	t.Setenv("MEMORY_EXTRACTOR_TIMEOUT_MS", "5000")
	if got := e.overrideTimeout(); got != 5*time.Second {
		t.Errorf("overrideTimeout() with env = %v, want 5s", got)
	}

	t.Setenv("MEMORY_EXTRACTOR_TIMEOUT_MS", "invalid")
	if got := e.overrideTimeout(); got != 0 {
		t.Errorf("overrideTimeout() with invalid env = %v, want 0", got)
	}
}

func TestEnvOrDefault(t *testing.T) {
	key := fmt.Sprintf("MEMORY_TEST_%d", time.Now().UnixNano())

	if got := envOrDefault(key, "fallback"); got != "fallback" {
		t.Errorf("envOrDefault unset = %q, want %q", got, "fallback")
	}

	t.Setenv(key, "custom")
	if got := envOrDefault(key, "fallback"); got != "custom" {
		t.Errorf("envOrDefault set = %q, want %q", got, "custom")
	}
}

// Compile-time verification that VideoExtractor implements Extractor.
var _ Extractor = (*VideoExtractor)(nil)
